### PR TITLE
Crypto: Rework device list tracking logic

### PR DESCRIPTION
**Pull request by ylecollen**
Equivalent to matrix-org/matrix-js-sdk#425.
Equivalent to matrix-org/matrix-js-sdk#431.

### CODE REVIEW

File: matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXDeviceList.java [hunk 1]
```diff
@@ -193,33 +184,83 @@ private void clearUnavailableServersList() {
...
+            for (String userId : userIds) {
+                if (!deviceTrackingStatuses.containsKey(userId)) {
+                    Log.d(LOG_TAG, "## startTrackingDeviceList() : Now tracking device list for " + userId);
+                    deviceTrackingStatuses.put(userId, TRACKING_STATUS_PENDING_DOWNLOAD);
+                    isUpdated = true;
+                }
+            }
+
+            if (isUpdated) {
+                mCryptoStore.saveDeviceTrackingStatuses(deviceTrackingStatuses);
```
Comment (manuroe) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXDeviceList.java]: This needs to be tested on device. The risk is to overload the store IO.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXDeviceList.java [hunk 2]
```diff
@@ -327,19 +380,17 @@ public void downloadKeys(List<String> userIds, boolean forceDownload, final ApiC
...
-
-                for (String userId : userIds) {
+        final List<String> downloadUsers = new ArrayList<>();
 
-                    Map<String, MXDeviceInfo> devices = store.getUserDevices(userId);
+        if (null != userIds) {
+            for (String userId : userIds) {
+                if (forceDownload) {
+                    Integer trackingStatus = mCryptoStore.getDeviceTrackingStatus(userId, TRACKING_STATUS_UNDEFINED);
+                    if (TRACKING_STATUS_UP_TO_DATE != trackingStatus) {
```
Comment (manuroe) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXDeviceList.java]: in case of forceDownload, we should do a download with no more checking.

Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXDeviceList.java]: fixed

File: matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXDeviceList.java (hunk 1 above)
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/crypto/MXDeviceList.java]: I tested on a device, it seems working properly.

**Pull request merged**